File: conformance/tests/oversize-bigint-shift/OversizeBigintShift.runar.py
from runar import (
    StatefulSmartContract, Bigint, public,
)


class OversizeBigintShift(StatefulSmartContract):
    """A length-sensitive byte op whose operand is a bigint const too large for
    a JSON number (NEW-008). ``9007199254740993`` is 2**53 + 1, one past
    ``Number.MAX_SAFE_INTEGER``, so the conformance runner cannot narrow it
    back to a JSON number and the checked-in ``expected-ir.json`` really does
    pin the ``"<n>n"`` STRING form that ``runar-cli compile`` writes to disk.
    On chain the operand is the 7-byte script number ``01 00 00 00 00 00 20``,
    ``OP_LSHIFT`` is length-preserving, and ``OP_INVERT`` flips all seven
    bytes."""

    out: Bigint

    def __init__(self, out: Bigint):
        super().__init__(out)
        self.out = out

    @public
    def shift(self):
        """Write the inverted, shifted oversize constant to state."""
        self.out = ~(9007199254740993 << 8)

File: conformance/tests/oversize-bigint-shift/OversizeBigintShift.runar.java
package runar.conformance.oversizebigintshift;

import runar.lang.StatefulSmartContract;
import runar.lang.annotations.Public;
import runar.lang.types.Bigint;

/**
 * OversizeBigintShift -- a length-sensitive byte op whose operand is a bigint
 * const too large for a JSON number (NEW-008). {@code 9007199254740993} is
 * 2^53 + 1, one past {@code Number.MAX_SAFE_INTEGER}, so the conformance
 * runner cannot narrow it back to a JSON number and the checked-in
 * {@code expected-ir.json} really does pin the {@code "<n>n"} STRING form that
 * {@code runar-cli compile} writes to disk. On chain the operand is the 7-byte
 * script number {@code 01 00 00 00 00 00 20}, {@code OP_LSHIFT} is
 * length-preserving, and {@code OP_INVERT} flips all seven bytes.
 */
class OversizeBigintShift extends StatefulSmartContract {

    Bigint out;

    OversizeBigintShift(Bigint out) {
        super(out);
        this.out = out;
    }

    /** Writes the inverted, shifted oversize constant to state. */
    @Public
    void shift() {
        this.out = Bigint.of(9007199254740993L).shl(Bigint.of(8)).not();
    }
}
